### PR TITLE
Add working nuxt example

### PR DESCRIPTION
**Pull request by lorado**
So here we go. 
As first - I am really very very sorry about this one big commit... I made multiple separated commits, hoping you could dig into each one and understand more where are some parts defined, but I made all commits to the new git repo that was created as I made new project... And I didn't figure out how I could move file history to the main repo... sorry for that.

Also - I am personally using yarn, so please also use yarn, because mixing yarn with npm may cause unexpected bugs. 

What I did:
- installed fresh nuxt via `yarn create nuxt-app` (currently version is able to install almost working TS setup, earlier you had to put all the things together by yourself)  
- tweaked last steps for fully TS support
- created `plugins/apollo.ts` Plugin
- ported and updated two pages from your vue example (`index` and `add-page`)
- configured graphql-codegen. To run it simply execute `yarn graphql-codegen`

Some disclaimers/infos:
- I was tweaking on apollo plugin about a year ago. It was kinda hard, because I like to use apollo in `asyncData` nuxt hook, that is executed on the server, therefore is renders completely full content giving you 100% SEO compatibility. Apollo uses `fetch` to create a HTTP request, but `fetch` doesn't exist in node. There is a `cross-fetch` library, that is exactly for this case, BUT I don't remember why, I was not able to get it running. Probably because of upload progress, as my project was relying on this feature.
So... if you have time and want to dig a little, you are welcome to try to marry apollo with cross-fetch.
I ended up using axios to imitate fetch API, because axios is anyway in the vue dependencies.
- I don't remember why exactly, probably of messing up with `fetch`, but newer version of apollo-client was not working for me, therefore I am currently using fixed version 2.6.8 in my project (latest is 2.6.10). I bumped apollo-client to the latest version, and it seems to work, I don't remember, probably I had a trouble with 2.6.9.  
-  In my project I am currently using JWT, therefore you will find some places where token gets attached to the request. As it works similar to paseto, I let it there, you have to adopt it
- There is also a token renew logic, you have to adopt it or drop it, depending on how paseto should work
- You have to choose some storage, that is available on both - on client and on server (SSR). Funny fact - I was using cookie to store the token. I guess you need to google for possible solutions
- on `add-page` you can see how upload progress listener is implemented. You can slow down your network connection on chrome and select big file to see the upload-progress in work ;)

And a workflow tip:
I configured graphql-codegen to do multiple things: 
- generate `schema.graphql` - this is useful for so called Developer Experience. Install a graphql-plugin for your IDE, and you will get autocompletion in you `.graphql`/`.gql` mutation/query files. That is really helpful 
- create types for any query/mutation in `graphql` folder - this is actually the primary task of graphql-codegen. All types are stored in `graphql/GQLTypes.ts`
- generate `fragmentTypes.json` - apollo need this file if you use union types in your schema.
So any time you change schema on backend, run `yarn graphql-codegen` to get IDE autocompletion for new fields. 
Any time you write new Query/Mutation - run `yarn graphql-codegen` to get Typescript types, which you then can use in your code. 

So go and try it out =) 
Before you start, read readme - you have to create a `nuxt.runtimeConfig.js` file. This is my solution to handle config. So it is possible to use different file on the server e.g., where environment is different.

### CODE REVIEW

**Comment by wolfiton** [comment]
Thanks for the detailed explanation I will have a go at it on Monday and see how everything fits and work.

Also, this might help with cross fetch problem https://github.com/apollographql/apollo-link/issues/513#issuecomment-566542940

**Comment by lorado** [comment]
Next comment is way better :D
https://github.com/apollographql/apollo-link/issues/513#issuecomment-580990233

As I said, try to use it with cross-fetch. It respects fetch specs. I mean isomorphic fetch didn't work for me either.

**Comment by wolfiton** [comment]
Thanks for the whole code I understood most of it but have some questions:

Shouldn't the refresh token be handled by apollo in-memory cache?

What does it happen to the app once there is no communication with the server? Does PWA features step in or just 500 error?

Shouldn't the login logic be handled in a middleware?

Thanks in advance

**Comment by lorado** [comment]
> Shouldn't the refresh token be handled by apollo in-memory cache?

Apollo cache is only the thing, that caches requests results, and it do it in a structured way. So if you query e.g. returns [Post] results, and you later do a query for a single Post with a subset of fields you did quering [Post], then apollo ist smart enough to return you a cached Post entity, without querying the server at all.

You have the ability to manipulate this cache manually, but this is most painful... E.g. if you make a mutation that creates a Post, you have to update the cache, and tell the cache, that this new Post entity should also be returned by querying `getPosts` query. Also, if you have something like `user { posts }`, where you query all posts from user, you have surely to update cache in that place. In my apps I currently simply clear the whole cache on every mutation, as this is the simplest solution.

In-Memory cache is just a one of many cache implementations for apollo, the simplest one. I never tried a different one. 

> Shouldn't the login logic be handled in a middleware?

Login means - you send a login + password to the server, and you become the token and store in any storage you like. In middleware you then simply append this token to the HTTP query apollo makes.

**Comment by wolfiton** [comment]
I found this which I am inclined to try.
https://www.youtube.com/watch?v=9knwu87IfU8
I don't know if you saw it, but if you didn't than make some time to watch it. I guarantee that it will not be time wasted because the presentation is made by a senior from GitLab that is specialized in Vue apollo and also contributed to the Vue Apollo core.

**Comment by lorado** [comment]
In my opinion - even if the god of all technologies would hold a talk about some technology or new way to use them, it doesn't means it fit in your project. In the talk you sent, they try to use only Apollo Cache instead of Apollo Cache + Vuex, as Apollo Cache meant to be a state, same as Vuex. Please note the terminology - Apollo Cache. InMemoryCache is only one possible implementation of Apollo Cache, there is surely more if them.

But back to the talk theme: go away from two states to single one. The idea is brilliant, as you become only one source of truth. 1,5 years ago I was on a local meetup, where some guy had the same idea, and he made his own implementation of Apollo Cache, that uses Redux to save the state (he was using react). So the opposite direction of what Natalia did. And it worked for him, kinda. 

In my opinion it is way simpler and cleaner to separate these two states, and let them work in their own territory: Vuex for App-State management, and Apollo Cache for graphql queries. I never store result of graphql query in Vuex, as it makes no sense to me. It maybe was a good thing to cache API calls earlier. I better simply query a graphql again, and if Apollo Cache has something for me, it serves it, without quering a server. And I am fine with it.
Also I think it is less verbose and less confusing to use Vuex, instead of extending a scheme on the client side. It sounds a little hacky, to write a custom resolver on the client for a field, that doesn't exists on the server...
For your example project this would be an overkill anyway. It is simpler to use Vuex, instead of trying to do the work in Apollo Cache/Apollo State. 

And trust me, manipulation of Apollo Cache - is a pain when it comes to adding or removing new entities. For example - in my first project I ever used GraphQL, we had multiple places, where `Post` may be queried, like:
- `posts` - returns all posts (you could filter them via input args)
- `user { posts }` - returns all posts of user
- `post { related_posts }` - returns all posts related to a given post.

Now for example you click through the app, and triggers queries for all listed possibilities. Now Apollo Cache have all results cached. For such fields Apollo Cache stores a list of entity IDs, that are in the list.
So now you go to the admin page, and create a new post, that is also marked as related to one post, that is in your Apollo Cache. After you made a mutation `createPost`, you have to update Apollo Cache. You have to append ID of the newly created Post to `posts`, `user.posts` and `post.related_posts` lists. If you don't do it, you will not see just created post, as apollo will never request the server, and your Apollo Cache has old cached data. Alone this modification of cache in the way I described - is very verbose.
Now image, later in your backend you introduce new field `dashboard.posts`, that returns all posts for your dashboard. I know, the example is not very representative, but imagine it would be. Now, you have to go to the places where you add new post, and update you code, because after mutation you have to update Apollo Cache for `dashboard.posts` field too! And this is a thing - you will probably forget to update it (as we did), and your app seems to be broken, because after adding a post, it doesn't show up in your dashboard. Only after page-reload it is there. So this sucks! 
That is why we decided to simply clean the whole Apollo Cache on ANY mutation. Yes, this removes the advantage of cache, but for us - we don't care, as users, that don't do any mutation and only do queries, they take advantage of the Apollo Cache, and that is fine for us.
And now - If you would store your app state in Apollo Cache, you will not have the ability to clean the whole cache, as this would also clean your app state, therefore you will need to manually update Apollo Cache after each mutation. Then you face cons of this approach I listed before.

To summarize - I think for your example app this would be simpler to use Vuex, as today this is the way to go to handle app state, and simply clean Apollo Cache on mutations.

Anyway I would like to know what @spawnia thinks about it

**Comment by spawnia** [comment]
I can only speak towards React/Apollo. I also found it quite difficult to store client cache in Apollo, so we went with a separate application state in the form of Redux.

**Comment by lorado** [comment]
And how do you handle Apollo Cache updates on mutations?

**Comment by spawnia** [comment]
Not great actually, I think everyone on our team does it differently: refetching queries, using network-only for some, actual cache updates, hard reloads, ... 🤷‍♂️ Haven't looked into it too much.

**Comment by wolfiton** [comment]
Maybe version 3 of apollo client will solve some problems https://www.apollographql.com/docs/react/data/queries/#caching-query-results

It would be wonderfull to be like this in vue right now.

**Comment by lorado** [comment]
This is not really a problem. This is apollo design of the graphql client.
You can see here the example of mutations that updates cache: https://apollo.vuejs.org/guide/apollo/mutations.html#server-side-example
And in general you can take a look at vue apollo docs: https://apollo.vuejs.org/guide/apollo/

**Comment by wolfiton** [comment]
Thanks maybe you can share this problem with apollo and also the vue apollo comunity and find a better solution @lorado.

That being said i have one question would you think is beter to use Options API (vanilla) or class based typescript(like in angular 2-9)?

This article should explain better what I am asking https://blog.logrocket.com/how-to-set-up-and-code-nuxt-js-apps-fully-in-typescript/.

Because i would like to also test my code after I finish it with jest(for nuxt) and pest for php

**Comment by lorado** [comment]
>Thanks maybe you can share this problem with apollo and also the vue apollo comunity and find a better solution @lorado.

Oh well... It works for me fine, therefore I don't have a reason to change there anything =) I have much more business driven problems to solve.

> That being said i have one question would you think is beter to use Options API (vanilla) or class based typescript(like in angular 2-9)?

Use whatever you like. I came to Vue from Angular, therefore I like this class based style. Vue 3 will have officially Option + Composition APIs. So you are free to choose, what you like. This is opinionated. As we currently have class based API in TS, this will definitely be updated to support vue 3, so... I mean most of the articles and documentations for Vue use Option API.

**Comment by spawnia** [comment]
> Because i would like to also test my code after I finish it with jest(for nuxt) and pest for php

My typical workflow is to write tests first or while I develop. That way I ensure the code is actually testable and works right away.

**Comment by wolfiton** [comment]
> > Because i would like to also test my code after I finish it with jest(for nuxt) and pest for php
> 
> Great idea. My typical workflow is to write tests first or while I develop. That way I ensure the code is actually testable and works right away.

I would agree to do TDD but unfortunately in this situation, I haven't built previously an application like this so, TDD might prove problematic.

**Comment by wolfiton** [comment]
Just wanted to let you guys know that today, I can't work on this example something came up, so i will try tomorrow to see how much time I have.

Thanks

**Comment by wolfiton** [comment]
I am following this setup for jest witrh typescript i hope it will work https://medium.com/@Al_un/nuxt-vuex-jest-tested-powered-by-typescript-70441600ef39

**Comment by wolfiton** [comment]
For pest I am using this guide https://nunomaduro.com/using-pest-in-laravel/

**Comment by wolfiton** [comment]
Should the user table include the roles or should I use two separated tables with one-to-many relationship?

**Comment by wolfiton** [comment]
Also i can use some database schema, this is what i was thinking:

Table Users with:
- username
- password

Table roles:
- name
- description

Table Ability:
- name
- description

Table ability_role many to many

Table role_user many to many


Table Post:
slug
title
content
cover_image
content_images maybe a many to many with  image_post pivot table
category_id


Table categories:
name
slug


Table tags:
name
slug

Table post_tag many to many

Any ideas are appreciated.

Thanks

**Comment by lorado** [comment]
>Should the user table include the roles or should I use two separated tables with one-to-many relationship?
It depends on your goals. If you want the ability to append multiple roles for a user, then you should create an extra table, where you store `user_id` and `role` (as enum value) this user has. 

In my project I tried something different - as I use Postgres, I decided to use Array-column. But for this I had to implement a custom class to cast this value. So for simplicity use an extra table =)

**Comment by lorado** [comment]
> Table roles:
> 
> * name
> * description
> 
> Table Ability:
> 
> * name
> * description
> 
> Table ability_role many to many
> 
> Table role_user many to many

This implies, that you can manage whole roles and abilities via DB. For the simplicity I would probably use only a role, that can be appended to a user. Role can also be a hard-coded enum, with possible values like "ADMIN", "MODERATOR" etc.
You can semantically treat "roles" as "permissions". The real abilities, what every role is able to do, you can define in [Policies](https://laravel.com/docs/7.x/authorization#creating-policies), and in GraphQL schema in laravel you can use `@can` directive, to call gates from policy.

**Comment by wolfiton** [comment]
I think that I need to do a jwt auth because, paseto right now doesn't make sense to me. I need to understand the flow and how laravel requests interacts with laravel lighthouse and if these 2 tehnologies need to be combined to send data to Stripe for example.

Becuase i don't seem to understand how I can use graphql to mimic what i do with simple laravel requests.

**Comment by lorado** [comment]
> I think that I need to do a jwt auth because, paseto right now doesn't make sense to me. I need to understand the flow and how laravel requests interacts with laravel lighthouse and if these 2 tehnologies need to be combined to send data to Stripe for example.

It doesn't metter, which technology you choose for authentication. JWT, Paseto, Session, whatever... You have to achieve one "simple" goal - fetch User, who is doing the request. This can be achieved due parsing a token from Bearer HTTP Header, or whatever you choose. When you then has a User model, you can use laravels Gates and Policies, to check, if user is authorized to do something

I don't get what you mean with "how laravel requests interacts with laravel lighthouse". Laravel Lighthouse adds one rote to laravel - `/graphql`. All requests sent to `/graphql` are treated by lighthouse. Request should be a valid GraphQL Request, what is done by apollo on the client side. And then depending on what you are sending - query or mutation - particular field resolver will be executed.

> Becuase i don't seem to understand how I can use graphql to mimic what i do with simple laravel requests.
I already implemented a custom simple mutation field resolver - https://github.com/wolfiton/vue-upload/blob/master/app/GraphQL/Mutations/CreatePostManually.php

Please, read lighthouse documentation, and try to implement features you need on your own, how you understand it. But please, try to do something, before you overthink every step 100 times... It is simpler to improve code later, when you have some base. Trust me, you will learn by doing. And if you get stuck, because you implemented something, and it doesn't do what you want, then you can ask again.

**Comment by wolfiton** [comment]
Hi,
So for the past several days, I learned more about the Laravel ecosystem so that I can create this project.

As a short summary I learned about:

Mail
Notifications
Events 
Listeners
Eloquent
Csrf
Authorization(still learning)
Creating simpler methods using the model
Some stripe stuff

https://laracasts.com/series/laravel-6-from-scratch -very useful especially using models to simplify code and also using validations directly for creation.

I will come back to it once I study more.

Thanks for the patience

**Comment by lorado** [comment]
Yesss, well done. You are on the right way. This is important to understand how laravel works 👍

**Comment by wolfiton** [comment]
Tomorrow I want to start with this:

- https://lighthouse-php-auth.com/
- https://github.com/vuex-orm/vuex-orm

My question is does lighthouse respect this adapters that vuex-orm-graphql has?

https://vuex-orm.github.io/plugin-graphql/guide/#adapters

**Comment by lorado** [comment]
I didn't use Laravel Passport till now. Dunno if it is good or bad.
Vuex-ORM sounds good, but it is for me an overload... I don't like to use too much libraries, that abstract everything so high. You can try it, but I don't think, that this is a typical library to use. If I understood it right, vuex-orm-graphql plugin generates queries/mutations automatically, so it seems like you don't need to define your own `.gql` files, therefore you lose advantage of graphql-codegen library.
Also this vuex-orm-graphql is way too much magic for me. The abstration level os so high, that it may become difficult to debug something, if something goes wrong. 
You are free to choose what you like. But after my experience with meteor, I will never use that high abstraction... 

Again, it seems to me, that you overthink way too much... I thought it will become a simple example, with simple setup, without any abstraction. Anyway, you can do what you like, but if you choose this way, I would not be able to help you further, because I am not familiar with these libraries

**Comment by wolfiton** [comment]
> I didn't use Laravel Passport till now. Dunno if it is good or bad.
> Vuex-ORM sounds good, but it is for me an overload... I don't like to use too much libraries, that abstract everything so high. You can try it, but I don't think, that this is a typical library to use. If I understood it right, vuex-orm-graphql plugin generates queries/mutations automatically, so it seems like you don't need to define your own `.gql` files, therefore you lose advantage of graphql-codegen library.
> Also this vuex-orm-graphql is way too much magic for me. The abstration level os so high, that it may become difficult to debug something, if something goes wrong.
> You are free to choose what you like. But after my experience with meteor, I will never use that high abstraction...
> 
> Again, it seems to me, that you overthink everything way too much... I thought it will become a simple example, with simple setup, without any abstraction. Anyway, you can do what you like, but if you choose this way, I would not be able to help you further, because I am not familiar with these libraries

I understand your position. But because  I understand how to use certain parts of vue. I never managed to find a good or clear tutorial on vuex. The only tutorial I found was a counter that really doidn't help a lot.

But As you said abstraction creates problems so tomorrow i will start simple and create the login and resgiter mutations. Do i have to create the jwt on my own or you could recommend a laravel package for that?

Thanks

**Comment by lorado** [comment]
TBO I currently use not the optimal package. At the moment jwt-auth is implemented: https://github.com/tymondesigns/jwt-auth 
This library has partially poor and partially outdated documentation. Some solutions I found in issues... But earlier (2 year ago) this was actually a single option to setup quick JWT auth with laravel, so that is why I did it. It works kinda for now, but we want to rework it soon. 

Today I would like to use sanctum: https://laravel.com/docs/7.x/sanctum
This is an official package, so I have much more confidence in it
There are also some people already using sanctum with lighthouse, so you can definitely get some help on lighthouse-slack channel: https://join.slack.com/t/lighthouse-php/shared_invite/enQtMzc1NzQwNTUxMjk3LWMyZWRiNWFmZGUxZmRlNDJkMTQ2ZDA1NzQ1YjVkNTdmNWE1OTUyZjZiN2I2ZGQxNTNiZTZiY2JlNmY2MGUyNTQ

Some words to vuex - in general you have 4 parts:
- **state** - this is actually your read only data. to change it, you have to use mutation
- **mutations** - simple functions, that can change state value. You have to use mutation to change state, so your state stays reactive. Vuex will anyway yell at you, if you try to change state outside of mutation. You are also not able to use any async stuff like promises or async/await. Thinks of mutations really like functions to define state value. 
- **actions** - async functions, where you can e.g. make a API request to the server, and also after getting request call mutation to update state
- **getters** - helper function that "prepares" your state. e.g. if you have in your state `firstname` and `lastname`, but you often need `fullname`, you can define a getter `fullname => firstname + ' ' + lastname`, so it is easier/cleaner to use it in your code.

You can take a look at this tutorial: https://scotch.io/tutorials/handling-authentication-in-vue-using-vuex
It is chopped in very small peaces. I hope you can understand auth process and how to use vuex with it a little more.

**Comment by wolfiton** [comment]
So I created my first test mutation using tdd but Pest doesn't recognize the `graphQL`function from `use Nuwave\Lighthouse\Testing\MakesGraphQLRequests;`

I even tried `MakesGraphQLRequests::graphQL`

I get the following 

```php
• Tests\Feature\Models\RegisterUserTest > test_user_register
   ErrorException 

  Undefined property: P\Tests\Feature\Models\RegisterUserTest::$MakesGraphQLRequests

  at tests/Feature/Models/RegisterUserTest.php:36
     32▕ 
     33▕     $password = "password"; //default for factories
     34▕ 
     35▕     $this
  ➜  36▕         ->MakesGraphQLRequests::graphQL(
     37▕             /** @lang GraphQL */
     38▕             '
     39▕             mutation ($name: String!, $username: String!, $email: String!, $password: String!) {
     40▕                 createUser(input: {

  1   tests/Feature/Models/RegisterUserTest.php:36
      Illuminate\Foundation\Bootstrap\HandleExceptions::handleError()

```

Also thanks for the vuex link it really helped a lot undestanding how i can reuse what I write in a vue component.

Thanks

**Comment by spawnia** [comment]
You have to call `uses(MakesGraphQLRequests::class)` somewhere before your test runs.

**Comment by wolfiton** [comment]
Found this whick will make our lives easier i think https://www.toptal.com/graphql/laravel-graphql-server-tutorial

**Comment by wolfiton** [comment]
Finished the tutorial and managed to convert it to the latest version 4.16 but at the end I run into a big problem that I don't know how to fix it

```php
type Mutation {
  # ...

  createArticle(title: String!, content: String!): Article 
    @field(resolver: "ArticleMutator@create")
    @middleware(checks: ["auth:api"])
}
```
First, let’s generate a mutation class:

```$ php artisan lighthouse:mutation ArticleMutator```

Next, let’s update the mutator with the following logic:
```php
namespace App\GraphQL\Mutations;

use Nuwave\Lighthouse\Support\Contracts\GraphQLContext;

class ArticleMutator
{
    /**
     * Return a value for the field.
     *
     * @param  null  $rootValue
     * @param  mixed[]  $args
     * @param  \Nuwave\Lighthouse\Support\Contracts\GraphQLContext  $context 
     * @return mixed
     */
    public function create($rootValue, array $args, GraphQLContext $context)
    {
        $article = new \App\Article($args);
        $context->user()->articles()->save($article);

        return $article;
    }
}
```
Note: We renamed the default resolve function to create. You don’t need to create a new class for every resolver. Instead, you can group logic together if it makes more sense.

Finally, let’s run our new mutation and check the output. Be sure to keep the Authorization header from our previous query in the “HTTP Headers” tab:
```gql
mutation {
  createArticle(
    title:"Building a GraphQL Server with Laravel"
    content:"In case you're not currently familiar with it, GraphQL is a query language used to interact with your API..."
  ) {
    id
    author {
      id
      email
    }
  }
}

```
**Error**
```php
{
  "errors": [
    {
      "debugMessage": "SQLSTATE[42703]: Undefined column: 7 ERROR:  column \"directive\" of relation \"articles\" does not exist\nLINE 1: insert into \"articles\" (\"title\", \"content\", \"directive\", \"us...\n                                                    ^ (SQL: insert into \"articles\" (\"title\", \"content\", \"directive\", \"user_id\", \"updated_at\", \"created_at\") values (Building a GraphQL Server with Laravel, In case you're not currently familiar with it, GraphQL is a query language used to interact with your API..., ?, 1, 2020-08-18 12:21:50, 2020-08-18 12:21:50) returning \"id\")",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 6,
          "column": 3
        }
      ],
      "path": [
        "createArticle"
      ],
```

**Comment by wolfiton** [comment]
What directive can I use instead of @middleware(checks: ["auth:api"]) to get the current user and adding it to the articles mutation?

Thanks

**Comment by spawnia** [comment]
I recommend you read the entire docs at https://lighthouse-php.com/ from start to finish.

**Comment by wolfiton** [comment]
I tried to use @inject directive but it didn't work 

If you wnat to look over my code and maybe guide me on how to resolve this problem that would be great here is the code https://github.com/wolfiton/laravel-graphql

The problem is in the createArticle mutation

**Comment by wolfiton** [comment]
Here are the 2 files that are related to my problem

https://github.com/wolfiton/laravel-graphql/blob/master/graphql/schema.graphql

https://github.com/wolfiton/laravel-graphql/blob/master/app/GraphQL/Mutations/ArticleMutator.php

**Comment by spawnia** [comment]
Just dig into the Laravel & Lighthouse docs and try things. Me giving you a solution to such trivial tasks will not help you. You need to build something yourself to learn.

**Comment by wolfiton** [comment]
If the solution is to use guard on the entire mutation than that is not a solution just a complication. 
A real solution would be to have a directive like @auth('user') that returns an auth user and creates a relationship on the current mutation.

Also if the fact that the author of lighthouse is using an older version 3. something to do a tutorial doesn't concern you and you don't have the following question: 
- Why is he doing that?
Well..

Here is the link and the tutorial is not older than 3 months 

https://www.toptal.com/graphql/laravel-graphql-server-tutorial
